include/sdf/Polyline.hh

```diff
@@ -0,0 +1,99 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#ifndef SDF_POLYLINE_HH_
+#define SDF_POLYLINE_HH_
+
+#include <vector>
+
+#include <ignition/math/Vector2.hh>
+#include <sdf/Error.hh>
+#include <sdf/Element.hh>
+#include <sdf/sdf_config.h>
+
+namespace sdf
+{
+  // Inline bracket to help doxygen filtering.
+  inline namespace SDF_VERSION_NAMESPACE {
+  //
+
+  // Forward declare private data class.
+  class PolylinePrivate;
+
+  /// \brief Polyline represents a 2D path. Multiple polylines can be combined
+  class SDFORMAT_VISIBLE Polyline
+  {
+    /// \brief Constructor
+    public: Polyline();
+
+    /// \brief Copy constructor
+    /// \param[in] _polyline Polyline to copy.
+    public: Polyline(const Polyline &_polyline);
+
+    /// \brief Move constructor
+    /// \param[in] _polyline Polyline to move.
+    public: Polyline(Polyline &&_polyline) noexcept;
+
+    /// \brief Destructor
+    public: virtual ~Polyline();
+
+    /// \brief Assignment operator.
+    /// \param[in] _polyline The polyline to set values from.
+    /// \return *this
+    public: Polyline &operator=(const Polyline &_polyline);
+
+    /// \brief Move assignment operator.
+    /// \param[in] _polyline Polyline to move.
+    /// \return Reference to this.
+    public: Polyline &operator=(Polyline &&_polyline);
+
+    /// \brief Load the polyline geometry based on an element pointer.
+    /// This is *not* the usual entry point. Typical usage of the SDF DOM is
+    /// through the Root object.
+    /// \param[in] _sdf The SDF Element pointer
+    /// \return Errors, which is a vector of Error objects. Each Error includes
+    /// an error code and message. An empty vector indicates no error.
+    public: Errors Load(ElementPtr _sdf);
+
+    /// \brief Get the polyline's height in meters.
+    /// \return The height of the polyline in meters.
+    public: double Height() const;
+
+    /// \brief Set the polyline's height in meters.
+    /// \param[in] _height The height of the polyline in meters.
+    public: void SetHeight(const double _height);
+
+    /// \brief Get the polyline's points. Each point has 2D coordinates in
+    /// meters.
+    /// \return The polyline's points.
+    public: const std::vector<ignition::math::Vector2d> &Points() const;
+
+    /// \brief Add a point to the polyline.
+    /// \param[in] _point 2D point to add.
+    public: void AddPoint(const ignition::math::Vector2d &_point);
```

Comment:
this is useful for incremental construction, but it doesn't allow removing or modifying points. Should we also provide a `SetPoints(const std::vector<ignition::math::Vector2d>&)` method?

Reply:
Yeah I was looking for this kind of feedback before adding a bunch of methods. I was just catching up on the [DOM API](https://github.com/ignitionrobotics/sdf_tutorials/blob/9ca00b9bcd3d698ad9e0c986b4cdb065e25aaf73/developers/dom_api.md#current-state-of-mutability) documentation and it looks like things are still up in the air.

Do you think `SetPoints` is enough? Should I remove `AddPoint`? Should I add other functions, maybe `ClearPoints`?

Reply:
ok, now that I look more closely at our API, I see that we don't typically return a `std::vector` like this. Based on [Link.hh](https://github.com/ignitionrobotics/sdformat/blob/sdf12/include/sdf/Link.hh), I would expect the following API here:

~~~
    public: uint64_t PointCount() const;

    public: const ignition::math::Vector2d *PointByIndex(const uint64_t _index) const;

    public: ignition::math::Vector2d *PointByIndex(const uint64_t _index);

    public: bool AddPoint(const ignition::math::Vector2d &_collision);

    public: void ClearPoints();
~~~

Reply:
Sounds good, implemented that API in [b376f96](https://github.com/ignitionrobotics/sdformat/pull/1000/commits/b376f964e16c1fe1eaad9b8806e6f33fd2af9fc4)

Reply:
As I'm updating Gazebo to use the new API, I realize how inconvenient it is not to have the `const std::vector<...> &Points()` function. Would you be ok bringing it back? I'm not sure what's the motivation for hiding the vector from the public API and instead handling raw pointers. I think we could make better use of the containers provided by the standard, which provide their own APIs for iterating, appending, accessing elements, etc.

Reply:
See what changes were needed to use the new API here: a89c6a028e0b930323a49b7c5c7ac2dc0664e14c

The vector was nice because it matched the Common API for extruding the polylines, but that's not enough reason to drive the API design here.

Reply:
ok, well it looks like we've started using `std::vector` in the DOM API:

* https://github.com/ignitionrobotics/sdformat/blob/sdf12/include/sdf/Plugin.hh#L102

I think it's fine to have an API that returns an `std::vector`; please feel free to restore it. Sorry for the runaround.

Reply:
Done in [c8cafb6](https://github.com/ignitionrobotics/sdformat/pull/1000/commits/c8cafb6bf2a56258373dc3569f716bce0f463d12)